File: apps/api/src/oneepis_api/schemas/hospitalization.py
from __future__ import annotations

import uuid
from datetime import date, datetime

from pydantic import Field

from oneepis_api.models.hospitalization import HospitalBedStatus
from oneepis_api.schemas.clinical_record import ClinicalEncounterRead
from oneepis_api.schemas.common import APIModel
from oneepis_api.schemas.patient import PatientRead


class HospitalBedBase(APIModel):
    ward: str = Field(min_length=1, max_length=80)
    room: str = Field(min_length=1, max_length=80)
    bed_label: str = Field(min_length=1, max_length=80)
    status: HospitalBedStatus = HospitalBedStatus.AVAILABLE
    encounter_id: uuid.UUID | None = None
    notes: str | None = Field(default=None, max_length=240)


class HospitalBedCreate(HospitalBedBase):
    pass


class HospitalBedUpdate(APIModel):
    ward: str | None = Field(default=None, min_length=1, max_length=80)
    room: str | None = Field(default=None, min_length=1, max_length=80)
    bed_label: str | None = Field(default=None, min_length=1, max_length=80)
    status: HospitalBedStatus | None = None
    encounter_id: uuid.UUID | None = None
    notes: str | None = Field(default=None, max_length=240)


class HospitalBedRead(HospitalBedBase):
    id: uuid.UUID
    created_at: datetime
    updated_at: datetime


class HospitalizationBoardItem(APIModel):
    patient: PatientRead
    encounter: ClinicalEncounterRead
    bed: HospitalBedRead | None = None


class HospitalDailySheetBase(APIModel):
    sheet_date: date
    clinical_summary: str = Field(min_length=1, max_length=2000)
    overnight_events: str | None = Field(default=None, max_length=2000)
    active_plan: str | None = Field(default=None, max_length=2000)
    pending_tasks: str | None = Field(default=None, max_length=2000)
    safety_notes: str | None = Field(default=None, max_length=2000)


class HospitalDailySheetCreate(HospitalDailySheetBase):
    pass


class HospitalDailySheetUpdate(APIModel):
    sheet_date: date | None = None
    clinical_summary: str | None = Field(default=None, min_length=1, max_length=2000)
    overnight_events: str | None = Field(default=None, max_length=2000)
    active_plan: str | None = Field(default=None, max_length=2000)
    pending_tasks: str | None = Field(default=None, max_length=2000)
    safety_notes: str | None = Field(default=None, max_length=2000)


class HospitalDailySheetRead(HospitalDailySheetBase):
    id: uuid.UUID
    patient_id: uuid.UUID
    encounter_id: uuid.UUID
    created_by: str
    created_at: datetime
    updated_at: datetime

File: apps/web/src/lib/type-contracts/hospitalization.ts
import type { ClinicalEncounter } from "./clinical-record";
import type { Patient } from "./patient";

export type HospitalBedStatus = "available" | "occupied" | "cleaning" | "blocked";
export type HospitalDailySheetStatus = "draft" | "closed";
export type HospitalIndicationStatus = "draft" | "closed";
export type HospitalBed = {
  id: string;
  ward: string;
  room: string;
  bed_label: string;
  status: HospitalBedStatus;
  encounter_id?: string | null;
  notes?: string | null;
  created_at: string;
  updated_at: string;
};

export type HospitalBedCreate = {
  ward: string;
  room: string;
  bed_label: string;
  status?: HospitalBedStatus;
  encounter_id?: string | null;
  notes?: string | null;
};

export type HospitalBedUpdate = Partial<HospitalBedCreate>;

export type HospitalizationBoardItem = {
  patient: Patient;
  encounter: ClinicalEncounter;
  bed?: HospitalBed | null;
};

export type HospitalDailySheet = {
  id: string;
  patient_id: string;
  encounter_id: string;
  status: HospitalDailySheetStatus;
  sheet_date: string;
  clinical_summary: string;
  overnight_events?: string | null;
  active_plan?: string | null;
  pending_tasks?: string | null;
  safety_notes?: string | null;
  created_by: string;
  created_at: string;
  updated_at: string;
};

export type HospitalDailySheetCreate = {
  sheet_date: string;
  status?: HospitalDailySheetStatus;
  clinical_summary: string;
  overnight_events?: string | null;
  active_plan?: string | null;
  pending_tasks?: string | null;
  safety_notes?: string | null;
};

export type HospitalDailySheetUpdate = Partial<HospitalDailySheetCreate>;

export type HospitalIndication = {
  id: string;
  patient_id: string;
  encounter_id: string;
  status: HospitalIndicationStatus;
  indicated_at: string;
  title: string;
  indication_text: string;
  rationale?: string | null;
  safety_notes?: string | null;
  created_by: string;
  created_at: string;
  updated_at: string;
};

export type HospitalIndicationCreate = {
  status?: HospitalIndicationStatus;
  indicated_at: string;
  title: string;
  indication_text: string;
  rationale?: string | null;
  safety_notes?: string | null;
};

export type HospitalIndicationUpdate = Partial<HospitalIndicationCreate>;
